vision_evaluation/retrieval_evaluators.py

```diff
@@ -0,0 +1,159 @@
+        ValueError: Samplewise metrics are not available outside of multilabel classification.
+        """
+        assert average == 'samples'
+        if targets.shape[1] == 1:
+            targets = np.append(targets, np.zeros((targets.shape[0], 1)), axis=1)
+            predictions = np.append(predictions, np.zeros((predictions.shape[0], 1)), axis=1)
+        return sm.precision_score(targets, predictions, average=average)
+
+    def get_report(self):
+        return {self._get_id(): self.calculate_score(average='samples', filter_out_zero_tgt=False)}
+
+
+class RecallAtKEvaluator(RetrievalEvaluator):
+    """
+    Recall evaluator for both multi-class and multi-label classification
+    """
+
+    def __init__(self, k):
+        super(RecallAtKEvaluator, self).__init__(TopKPredictionFilter(k))
+        self.k = k
+
+    def _get_id(self):
+        return f'recall_at_{self.k}'
+
+    def _calculate(self, targets, predictions, average='samples'):
+        """
+        There is a special case that needs to be handled appropriately.
+        Due to filtering conditions, there are occasions where only 1 class remains in targets/predictions and sklearn interpretes this as an invalid configuration for multilabel.
+        Since when average == 'samples' is only calculated for multilabel, sm.recall_score throws an exception. e.g.:
+        predictions = np.array([[True]])
+        targets = np.array([[True]])
+        recall_score(targets, predictions, average='samples') throws the following error:
+        ValueError: Samplewise metrics are not available outside of multilabel classification.
+        """
+        assert average == 'samples'
+        if targets.shape[1] == 1:
+            targets = np.append(targets, np.zeros((targets.shape[0], 1)), axis=1)
+            predictions = np.append(predictions, np.zeros((predictions.shape[0], 1)), axis=1)
+        return sm.recall_score(targets, predictions, average=average)
+
+    def get_report(self):
+        return {self._get_id(): self.calculate_score(average='samples')}
+
+
+class MeanAveragePrecisionAtK(RetrievalEvaluator):
+    """
+    MeanAveragePrecision @ K as defined here:
+    https://stackoverflow.com/questions/54966320/mapk-computation
+    https://medium.com/@misty.mok/how-mean-average-precision-at-k-map-k-can-be-more-useful-than-other-evaluation-metrics-6881e0ee21a9
+    MAP@K is the average of AveP(K) over all queries. Hence it uses average="samples".
+    AveP(K) =  Num / Den, where:
+    Num = sum_i^k P(i) * rel(i), where:
+    P(i) is the Precision @ i
+    rel(i) is an indicator function where rel(i) = 1 if position i is relevant, and rel(i) = 0 otherwise.
+    Den = min(K, number of relevant images)
+    """
+
+    def __init__(self, k):
+        super(MeanAveragePrecisionAtK, self).__init__()
+        self.k = k
+
+    def _get_id(self):
+        return f'map_at_{self.k}'
+
+    def calculate_score(self):
```

Comment:
Same here

Reply:
let us not alter the arg list if it overrides the base class's method

Reply:
same here, calculate_score() is called by get_report() so why add unnecessary params? Am I missing something here?

Reply:
For anyone who works with changes in base class, they have to check all implemntations of sub-classes like this one now.

Reply:
A call to _calculate(targets, predictions, average) in MemorizingEverythingEvaluator will fail this class, even thought _calculate was originaly defined in MemorizingEverythingEvaluator, with targets, predictions, average as arguments.

Reply:
`
class C1:
    def __init__(self) -> None:
        pass

    def run(self):
        self.write_text('1', '2')

    def write_text(self, text, image):
        print(text)
        print(image)


class C2 (C1):
    def write_text(self):
        print('3')
        print('4')


c = C2()
C2.run()

`

Reply:
"""
class C1:
    def __init__(self) -> None:
        pass

    def run(self):
        self.write_text('1', '2')

    def write_text(self, text, image):
        print(text)
        print(image)


class C2 (C1):
    def write_text(self):
        print('3')
        print('4')


c = C2()
c.run()
"""

**>>> TypeError: write_text() takes 1 positional argument but 3 were given**